x-pack/plugins/apm/server/lib/transactions/distribution/get_buckets/index.ts

```diff
@@ -3,35 +3,204 @@
+    const response = await apmEventClient.search({
+      apm: {
+        events: [ProcessorEvent.transaction],
+      },
+      body: {
+        query: {
+          bool: {
+            filter: [
+              ...commonFilters,
+              { term: { [TRANSACTION_SAMPLED]: true } },
+            ],
+            should: [
+              { term: { [TRACE_ID]: traceId } },
+              { term: { [TRANSACTION_ID]: transactionId } },
+            ],
+          },
+        },
+        aggs: {
+          distribution: {
+            histogram: getHistogramAggOptions({
+              bucketSize,
+              field: TRANSACTION_DURATION,
+              distributionMax,
+            }),
+            aggs: {
+              samples: {
+                top_hits: {
+                  _source: [TRANSACTION_ID, TRACE_ID],
+                  size: 10,
+                  sort: {
+                    _score: 'desc',
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    });
+
+    return (
+      response.aggregations?.distribution.buckets.map((bucket) => {
+        return {
+          key: bucket.key,
+          samples: bucket.samples.hits.hits.map((hit) => ({
+            traceId: hit._source.trace.id,
+            transactionId: hit._source.transaction.id,
+          })),
+        };
+      }) ?? []
+    );
+  }
+
+  async function getDistributionBuckets() {
+    const response = await apmEventClient.search({
+      apm: {
+        events: [
+          getProcessorEventForAggregatedTransactions(
+            searchAggregatedTransactions
+          ),
+        ],
+      },
+      body: {
+        query: {
+          bool: {
+            filter: [
+              ...commonFilters,
+              ...getDocumentTypeFilterForAggregatedTransactions(
+                searchAggregatedTransactions
+              ),
+            ],
+          },
+        },
+        aggs: {
+          distribution: {
+            histogram: getHistogramAggOptions({
+              field: getTransactionDurationFieldForAggregatedTransactions(
+                searchAggregatedTransactions
+              ),
+              bucketSize,
+              distributionMax,
+            }),
+          },
+        },
+      },
+    });
+
+    return (
+      response.aggregations?.distribution.buckets.map((bucket) => {
+        return {
+          key: bucket.key,
+          count: bucket.doc_count,
+        };
+      }) ?? []
+    );
+  }
+
+  const [
+    samplesForDistributionBuckets,
+    distributionBuckets,
+  ] = await Promise.all([
+    getSamplesForDistributionBuckets(),
+    getDistributionBuckets(),
+  ]);
+
+  const buckets = joinByKey(
+    [...samplesForDistributionBuckets, ...distributionBuckets],
+    'key'
+  ).map((bucket) => ({
+    ...bucket,
+    samples: bucket.samples ?? [],
+    count: bucket.count ?? 0,
+  }));
+
+  return {
+    noHits: buckets.length === 0,
+    bucketSize,
+    buckets,
+  };
 }
```

Comment:
I see the "fetcher" and "transform" has been merged. I think the separation was nice in theory but perhaps it's not working out?

Reply:
It felt awkward moving back and forth between three files. As the fetch implementation was already significantly longer than the transform, I merged the three files into one.